clang/runtime/dpct-rt/include/math.hpp

```diff
@@ -242,6 +242,27 @@ inline unsigned unordered_compare_mask(const sycl::marray<T, 2> a,
       .as<sycl::vec<unsigned, 1>>();
 }
 
+template <typename T>
+inline T bfe(T source, uint32_t bit_start, uint32_t num_bits) {
+  if (!num_bits)
+    return 0;
+  bit_start &= 0xFF;
+  num_bits &= 0xFF;
+  const T mask = (T{1} << num_bits) - 1;
+  return (source >> bit_start) & mask;
+}
+
+template <typename T>
+inline T bfi(T x, T y, uint32_t bit_start, uint32_t num_bits) {
+  constexpr unsigned msb = std::numeric_limits<T>::digits;
+  bit_start &= 0xFF;
+  num_bits &= 0xFF;
+  if (bit_start > msb || num_bits == 0)
+    return y;
+  T mask = (~(T{}) >> (msb - num_bits)) << bit_start;
```

Comment:
Without knowing the input type, `T`, we are making assumptions about the default constructor.  Probably better to use `T{0}`?  

I also think it may be good to return the `::std::enable_if_t<::std::is_unsigned_v<T>, T>`  to both functions to ensure the incoming type makes sense.  If callers of this function want to reinterpret cast their input data, they can before calling.  If we did this, the assumptions about the default constructor are OK (but maybe more clear to be explicit anyway).

Reply:
> Without knowing the input type, T, we are making assumptions about the default constructor. Probably better to use T{0}?

Agree.
> I also think it may be good to return the ::std::enable_if_t<::std::is_unsigned_v<T>, T> to both functions to ensure the incoming type makes sense. If callers of this function want to reinterpret cast their input data, they can before calling. If we did this, the assumptions about the default constructor are OK (but maybe more clear to be explicit anyway).

For PTX bfe instruction, it allowed T as a signed integer type, so I relax constraints.

Reply:
done